File: 500.600/502.ipo.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class ipo
{
public:
    int findMaximizedCapital(int k, int w, const vector<int> &profits, const vector<int> &capital)
    {
        int n = profits.size();
        vector<pair<int, int>> arr;
        for (int i = 0; i < n; ++i)
        {
            arr.push_back({capital[i], profits[i]});
        }
        sort(arr.begin(), arr.end());
        int curr = 0;
        priority_queue<int, vector<int>, less<int>> pq;
        for (int i = 0; i < k; ++i)
        {
            while (curr < n && arr[curr].first <= w)
            {
                pq.push(arr[curr].second);
                ++curr;
            }
            if (!pq.empty())
            {
                w += pq.top();
                pq.pop();
            }
            else
            {
                break;
            }
        }
        return w;
    }
};

int main(int argc, char const *argv[])
{
    ipo cls;
    cout << cls.findMaximizedCapital(2, 0, vector<int>{1, 2, 3}, vector<int>{0, 1, 1}) << " -> 4" << endl;
    cout << cls.findMaximizedCapital(3, 0, vector<int>{1, 2, 3}, vector<int>{0, 1, 2}) << " -> 6" << endl;
    return 0;
}

File: go/500.600/502.ipo.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

type hp struct {
	sort.IntSlice
}

func (h *hp) Less(i int, j int) bool {
	return h.IntSlice[i] > h.IntSlice[j]
}

func (h *hp) Push(x any) {
	h.IntSlice = append(h.IntSlice, x.(int))
}

func (h *hp) Pop() any {
	a := h.IntSlice
	v := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return v
}

type pair struct {
	c, p int
}

func findMaximizedCapital(k int, w int, profits []int, capital []int) int {
	n := len(profits)
	arr := make([]*pair, n)
	for i, p := range profits {
		arr[i] = &pair{c: capital[i], p: p}
	}
	sort.Slice(arr, func(i, j int) bool {
		return arr[i].c < arr[j].c
	})
	h := &hp{}
	for cur := 0; k > 0; k-- {
		for cur < n && arr[cur].c <= w {
			heap.Push(h, arr[cur].p)
			cur++
		}
		if h.Len() == 0 {
			break
		}
		w += heap.Pop(h).(int)
	}
	return w
}

func main() {
	fmt.Println(findMaximizedCapital(2, 0, []int{1, 2, 3}, []int{0, 1, 1}), 4)
	fmt.Println(findMaximizedCapital(3, 0, []int{1, 2, 3}, []int{0, 1, 2}), 6)
}
